Enhance and add more tests for ResizeRequest

gaobinlong opened the pull request:
This PR relates to #65714 and follows #67705.

The main changes of this PR are:
1.  Add the missing `equals` and `hashCode` methods in `ResizeRequest`.
2. ~Add the newly added `max_single_primary_size` parameter to the `XContent` of `ResizeRequest `.~ (edit: pulled out into #68793)
3. Add more test methods such as `testEqualsAndHashCode` and `testSerializeRequest` for `ResizeRequest`.

Commit:
Enhance and add more tests for ResizeRequest

gaobinlong commented:
@dakrone ,can you take a look at this PR?

elasticmachine commented:
Pinging @elastic/es-core-features (Team:Core/Features)

dakrone commented:
@elasticmachine ok to test

dakrone commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix_ResizeRequest

joegallo commented on server/src/test/java/org/elasticsearch/action/admin/indices/shrink/ResizeRequestTests.java:
```diff
@@ -92,11 +105,79 @@ public void testToAndFromXContent() throws IOException {
+        ResizeRequest deserializedReq = new ResizeRequest(wrap);
+
+        assertEquals(request.getSourceIndex(), deserializedReq.getSourceIndex());
+        assertEquals(request.getTargetIndexRequest().settings(), deserializedReq.getTargetIndexRequest().settings());
```
```
assertEquals(request.getTargetIndexRequest(), deserializedReq.getTargetIndexRequest());
```

Adding ^ here would make `testSerializeRequest()` fail, and so I think that means the `equals` method isn't actually correct as written (well, the code of it is fine, but since `CreateIndexRequest` doesn't override `equals`, too, we still end up doing the comparison as identity rather than equivalence there.

joegallo replied:
I'm still looking into this -- it's a bit of a sticky situation. 😬

gaobinlong replied:
Can we just compare the `settings` and `alias` of the `targetIndexRequest` in the `equals` method?

joegallo replied:
I pulled some of this out into #68793 so that we can get it merged through before 7.12.0, but I'm still fussing with things on the `equals`/`hashCode` front. Thanks for your patience, @gaobinlong.

joegallo committed:
Make RandomCreateIndexGenerator easier to re-use

by converting it to unchecked exceptions, and then use it to create
random create index requests.

joegallo committed:
Rename createTestItem to createTestInstance

joegallo committed:
Add equals/hashCode to CreateIndexRequest

joegallo committed:
Use AbstractWireSerializingTestCase

to cover testEqualsAndHashcode and testSerialization

joegallo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into fix_ResizeRequest

joegallo committed:
Merge branch 'master' into fix_ResizeRequest

joegallo committed:
No need to manually testEqualsAndHashCode

This is handled for us by AbstractWireTestCase#testEqualsAndHashcode

joegallo commented:
@gaobinlong thanks again for your patience, and apologies that it took me so long to get back to this. I've added some new commits -- do you think this is still a fair attempt at what you were trying to accomplish?

gaobinlong commented:
Yeah, it looks great! That's really what I wanted!

joegallo commented:
@dakrone I don't think it's fair for me to be a reviewer here anymore -- do you mind taking a look at things?

dakrone reviewed (approved):
LGTM, thanks both of you for iterating on it
